### PR TITLE
PorousFlow allows units to be specified

### PR DESCRIPTION
**Pull request by WilkAndy**
Refs #13154

Requesting review by @cpgr - There are no tests yet, but i'd appreciate it if you could read the doco when it builds  (PorousFlowSingleComponentFluid.md) to see whether you agree with the overall strategy.

### CODE REVIEW

File: modules/porous_flow/doc/content/source/materials/PorousFlowSingleComponentFluid.md [hunk 1]
```diff
@@ -2,9 +2,149 @@
...
+
+In addition, you must remember that fluid and heat sources must be measured in kg.m$^{-3}$.dy$^{-1}$ and J.m$^{-3}$.dy$^{-1}$, respectively.  Remember too that the thermal conductivity of the solid skeleton must be measured in J.m$^{-1}$.K$^{-1}$.dy$^{-1}$.  Velocities, radioactive decay rates, chemical precipitation and reaction rates, dispersion tensorsand diffusion coefficients are similarly impacted.
+
+Remember that the FluidProperties module always uses Pa, m, s and J.  Therefore, when using [SimpleFluidProperties](SimpleFluidProperties.md) you must specify the fluid thermal conductivity and viscosity in these units, not using dy.
+
+
+### Years
+
+Choosing the time unit to be years means no change is made to the inputs to the FluidProperties module.  The viscosity returned by the FluidProperties module is multiplied by $1/(3600\times 24\times 365.25)$ (and any quantity from choosing non-default pressure units) before handing back to the remainder of PorousFlow.
```
Comment (cpgr) [inline modules/porous_flow/doc/content/source/materials/PorousFlowSingleComponentFluid.md]: Maybe a note that this it the Julian year definition?

File: modules/porous_flow/doc/content/source/materials/PorousFlowSingleComponentFluid.md [hunk 2]
```diff
@@ -2,9 +2,149 @@
...
+
+In addition, you must remember that fluid and heat sources must be measured in kg.m$^{-3}$.yr$^{-1}$ and J.m$^{-3}$.yr$^{-1}$, respectively.  Remember too that the thermal conductivity of the solid skeleton must be measured in J.m$^{-1}$.K$^{-1}$.yr$^{-1}$.  Velocities, radioactive decay rates, chemical precipitation and reaction rates, dispersion tensorsand diffusion coefficients are similarly impacted.
```
Comment (cpgr) [inline modules/porous_flow/doc/content/source/materials/PorousFlowSingleComponentFluid.md]: Missing space - tensorsand -> tensors and

File: modules/porous_flow/doc/content/source/materials/PorousFlowSingleComponentFluid.md [hunk 3]
```diff
@@ -2,9 +2,149 @@
...
+
+### Days
+
+Choosing the time unit to be days means no change is made to the inputs to the FluidProperties module.  The viscosity returned by the FluidProperties module is multiplied by $1/(3600\times 24)$ (and any quantity from choosing non-default pressure units) before handing back to the remainder of PorousFlow.
+
+In addition, you must remember that fluid and heat sources must be measured in kg.m$^{-3}$.dy$^{-1}$ and J.m$^{-3}$.dy$^{-1}$, respectively.  Remember too that the thermal conductivity of the solid skeleton must be measured in J.m$^{-1}$.K$^{-1}$.dy$^{-1}$.  Velocities, radioactive decay rates, chemical precipitation and reaction rates, dispersion tensorsand diffusion coefficients are similarly impacted.
+
+Remember that the FluidProperties module always uses Pa, m, s and J.  Therefore, when using [SimpleFluidProperties](SimpleFluidProperties.md) you must specify the fluid thermal conductivity and viscosity in these units, not using dy.
```
Comment (cpgr) [inline modules/porous_flow/doc/content/source/materials/PorousFlowSingleComponentFluid.md]: dy -> days?

File: modules/porous_flow/doc/content/source/materials/PorousFlowSingleComponentFluid.md [hunk 4]
```diff
@@ -2,9 +2,149 @@
...
+
+
+### Years
+
+Choosing the time unit to be years means no change is made to the inputs to the FluidProperties module.  The viscosity returned by the FluidProperties module is multiplied by $1/(3600\times 24\times 365.25)$ (and any quantity from choosing non-default pressure units) before handing back to the remainder of PorousFlow.
+
+In addition, you must remember that fluid and heat sources must be measured in kg.m$^{-3}$.yr$^{-1}$ and J.m$^{-3}$.yr$^{-1}$, respectively.  Remember too that the thermal conductivity of the solid skeleton must be measured in J.m$^{-1}$.K$^{-1}$.yr$^{-1}$.  Velocities, radioactive decay rates, chemical precipitation and reaction rates, dispersion tensorsand diffusion coefficients are similarly impacted.
+
+Remember that the FluidProperties module always uses Pa, m, s and J.  Therefore, when using [SimpleFluidProperties](SimpleFluidProperties.md) you must specify the fluid thermal conductivity and viscosity in these units, not using yr.
```
Comment (cpgr) [inline modules/porous_flow/doc/content/source/materials/PorousFlowSingleComponentFluid.md]: yr -> years?

File: modules/porous_flow/doc/content/source/materials/PorousFlowSingleComponentFluid.md [hunk 5]
```diff
@@ -2,9 +2,149 @@
...
+
+- For instance, suppose you choose to measure pressure in MPa.  Then `PorousFlowSingleComponentFluid` will multiply your pressure by $10^6$ before feeding it to the FluidProperties module.  Since the viscosity returned by the FluidProperties module has units Pa.s, it needs to be converted to MPa.s before using it in the remainder of PorousFlow, so the FluidProperties viscosity is multiplied by $10^{-6}$.
+
 !alert note
-The FluidProperties userobjects expect temperature in Kelvin. If the simulation uses temperature in
-Celsius, `temperature_units = celsius` must be used.
+If you choose non-default SI units, you should ensure that all the other parameters in your input file use these non-default units.  For instance, if you choose to measure pressure in MPa, then your solid-mechanical stresses, Young's moduli, strengths, etc, must also be specified in MPa.  That is, `PorousFlowSingleComponentFluid` only interfaces with the FluidProperties module: it doesn't also change your pressure boundary conditions, end time, gravity, etc: all these must be specified by you in the units you choose.
```
Comment (cpgr) [inline modules/porous_flow/doc/content/source/materials/PorousFlowSingleComponentFluid.md]: Maybe make it stronger like "you must ensure that"

And perhaps a note that PF cannot check consistency of units automatically?

**Review by cpgr — COMMENTED** [review]
Should you put a big warning up front saying that it is probably easier/ strongly recommended/ much less error prone to just use default SI units and change units with some diy  post processing?

**Comment by WilkAndy** [comment]
I get the impression that you think this is a dumb idea, @cpgr.  Eg, we're setting ourselves and the user up for a lot of trouble.   I think it's a good idea in principal but i don't see a good way of making it foolproof and easy to use.

**Comment by cpgr** [comment]
Haha - no, I think it is useful! I was just trying to make sure that the documentation stresses that we can't ensure consistency.

**Comment by cpgr** [comment]
Looks good Andy

**Comment by WilkAndy** [comment]
Thanks @cpgr

**Review by permcody — APPROVED** [review]

**Comment by WilkAndy** [comment]
Thanks for the review and merge @cpgr and @permcody

**Pull request merged**